### ISSUE TITLE
Protocol

### ISSUE THREAD

**Issue by deepakkt** [Nyr/openvpn-install#174]
In my place, UDP is blocked.

TCP goes through only 80, 443.
I was able to get your script to work by changing UDP to TCP.

Making it configurable in the script in addition to the other settings would be lovely.

**Comment by Nyr** [comment]
I understand your use case but adding an option for this is not planned, since very often TCP is used wrongly with OpenVPN.

For those really needing it, just make a quick edit in the configuration file ;)

**Comment by valery1707** [comment]
@deepakkt this option exists in [openvpn-install-advanced](https://github.com/zabullet/openvpn-install-advanced) by @zabullet

**Comment by deepakkt** [comment]
Oh ok. I was however able to get it to work by changing this script. Worked
right away! Will keep the other option in mind for the next deploy.

**Comment by Nyr** [comment]
Remember that unofficial forks aren't always well maintained.

They are more than welcome, but different maintainers have different support policies and standards.

**Comment by zabullet** [comment]
@Nyr FWIW I do try and track your changes (if you look at my commit logs) and bring them to the pl48415 branch.

**Comment by Nyr** [comment]
@zabullet interesting, they always mention your fork and I didn't notice it was a forked from @pl48415.

The problem from my point of view, is (among others) that people directed to your main branch is using an outdated script which lacks some important features.

Just an example, from the readme of the fork:

> You can create personal DNS resolver that is accessible only through vpn so you can tunnel even DNS requests.

This implies a dangerous misunderstanding of how DNS works over OpenVPN. My original work is already configured to tunnel internet traffic over the VPN. Connections to external DNS are routed over the secure connection by default.

Coincidentally the fork lacks acca10ba1a87b86da3f402ef08e07ad42126913e, which is a very important commit to prevent [DNS leaks on Windows 10](https://medium.com/@ValdikSS/beware-of-windows-10-dns-resolver-and-dns-leaks-5bc5bfb4e3f1).

The fact that the fork also includes an option to scan all HTTP traffic using an antivirus is... interesting to say at least.

Security is not offering larger key lengths, "stronger" cipher choices or a freakin' antivirus. That's why I prefer to keep discussion of forks advertising this separated from my main project.

**Comment by zabullet** [comment]
@Nyr fwiw I'm doing no self promotion with regard to that repo or my branch....seems I've got a couple of users who like it and are vocal in such regard. I've read a lot of your posts and followed your repo and respect your decision to not include UDP as part of your config options. That being said....there is clearly a demand for it....and I wouldn't expect you to compromise you reputation for a feature that doesn't meet the design goals of your script. 

I'll try and bring as many of your changes and [purist] security ethics to the code base I as I can....but will publicly state that it is "best effort" on my part.

I'd welcome a discussion of your thoughts on openvpn-install-advance script, but suspect you'll find it in conflict with your ideals.

In the time being I'll look at that commit and try and bring it in.

zb.

**Comment by Nyr** [comment]
Just a small clarification:

Even if I don't like the direction of someone else's work and am vocal about it, or even if I personally think that it has negative implications, it doesn't change the fact that my work is released under a very permissive license. Everyone is free to do as they please, no questions asked or permission, validation or whatever needed from me.

About openvpn-install-advanced in particular, since you asked:

I personally think that it advertises a false sense of security and I don't like that, I have criticized other forks for similar reasons so I doubt it's news for you or anyone else following this project ;) It could be interesting to see an "advanced" fork which adds popular features like TCP, multiple connections per certificate, cipher and key length choices, password authentication, DNSCrypt, internal routing, etc... there are lots of room for expansion, but it's also a bit difficult to maintain something like that while keeping a clean user experience and maintaining everything sane, compatible and up to date. There are single-line changes which I've literally researched, tested or troubleshooted for hours in the past. Error checking is another field where my work could see lots of improvement, although it's also very tedious to do it right.

There are many other things which I would like to see improved, but I need to make some compromises. Personally, I am in a... bad situation right now and I have fully abandoned all my internet presence, which is pretty disastrous for someone in the IT field. This is the only project that I am maintaining and plan to keep maintaining, but for anyone wondering I consider it mature enough at this moment and don't plan "big" changes for the near future.

I also considered an `--advanced` toogle with lots of well documented options in the past. Hopefully some day...

(Sorry, this got a bit off-topic)

**Comment by lukebarone** [comment]
@Nyr How about forking off a section *just* for an `--advanced` toggle, and
incorporate *some* of the ideas? Even if other people are doing the
testing/debugging? I have some Digital Ocean-hosted servers that I can test
various options on, as well as many client systems to use to ensure the
options work...

**Comment by Nyr** [comment]
@lukebarone I appreciate your offer but I'm not a fan of "works for me" testing for a small project like this, mainly because it doesn't have a big enough base of beta-testers. At the moment, every "significant" change is tested with different Debian, Ubuntu and CentOS versions on all of the following plataforms:

- Default OS images installed from the ISO
- Templates from DigitalOcean, Vultr and often other popular providers: they include significant modifications, like SELinux being disabled on CentOS.
- Common OpenVZ templates: some are very common but lack important tooks included by the default OS, others have NAT or weird setups, etc...

This has been the way I have managed to push stable releases over the years, not saying that it is the best but has been working. If you take a look at the list of commits you'll see that some of them are preventing rare issues which would had been unnoticed otherwise (7fb12dc5cbd08904a88f8eead1c75f43590f7457, cf60872eaef477c1510bc1f9881c29ed0b6fc57b, etc...)

To summarize:
- Bug-fixes and improvements are very welcome. New features not very likely to be merged.
- I'm very interested in maintaining the existing work and will keep doing it as always.
- Many of the users running this script want simplicity and some don't really understand what they are doing or how it's working. I also know that others really want advanced options for a more customized setup.
- "Advanced" configuration toggles can happen but not right now. I would want them to be clean, well documented and "hidden" from the main UX. If someone wants them right now, or would prefer a different approach, please start a fork.

I hope this clarifies my position a bit.

**Comment by acs-ferreira** [comment]
https://en.wikipedia.org/wiki/KISS_principle
